### PR TITLE
Add runtime support for `PreRuntime` and `Consensus` digests

### LINKED ISSUE #2738: `PreRuntime` support in runtimes
#2466 broke all existing runtimes, since none of them were prepared to handle `PreRuntime` digests.  This must be fixed.

~~Along the way, I plan to do a global `PreRuntime` → `Inherent` renaming across the codebase.~~

### PR DESCRIPTION
**Pull request by DemiMarie-parity**
Fixes 6e253c899e67ce4e29b70e5eca3a3763fa42233f
Fixes #2738

### CODE REVIEW

File: core/sr-primitives/src/lib.rs [hunk 1]
```diff
@@ -611,6 +616,62 @@ macro_rules! impl_outer_config {
 	}
 }
 
+#[macro_export]
+#[doc(hidden)]
+macro_rules! __parse_pattern_2 {
+	(PreRuntime $module:ident $internal:ident $v1:ident $v2:ident) => {
+		$internal::$module($module::RawLog::PreRuntime(ref $v1, ref $v2, $crate::PhantomData))
```
Comment (thiolliere) [inline core/sr-primitives/src/lib.rs]: one solution to avoid this special parse is to make a variant being a unique struct like this. (even if its boring to derive because of incorrect bound by standard derive macros)
```diff
diff --git a/srml/aura/src/lib.rs b/srml/aura/src/lib.rs
index 7cc7e2755..7cad009c2 100644
--- a/srml/aura/src/lib.rs
+++ b/srml/aura/src/lib.rs
@@ -102,13 +102,39 @@ pub type Log<T> = RawLog<T>;
 ///
 /// The type parameter distinguishes logs belonging to two different runtimes,
 /// which should not be mixed.
-#[cfg_attr(feature = "std", derive(Serialize, Debug))]
-#[derive(Encode, Decode, PartialEq, Eq, Clone)]
+#[cfg_attr(feature = "std", derive(Serialize))]
+#[derive(Encode, Decode)]
+#[serde(bound(serialize = "T: Trait"))]
 pub enum RawLog<T> {
        /// AuRa inherent digests
-       PreRuntime([u8; 4], Vec<u8>, PhantomData<T>),
+       PreRuntime(PreRuntime<T>),
+}
+
+impl<T> std::fmt::Debug for RawLog<T> {
+       fn fmt(&self, f: &mut std::fmt::Formatter) -> Result<(), std::fmt::Error> {
+               unimplemented!();
+       }
+}
+
+impl<T> Clone for RawLog<T> {
+       fn clone(&self) -> Self {
+               unimplemented!();
+       }
 }
 
+impl<T> Eq for RawLog<T> {
+}
+
+impl<T> PartialEq for RawLog<T> {
+       fn eq(&self, other: &Self) -> bool {
+               unimplemented!();
+       }
+}
+
+#[cfg_attr(feature = "std", derive(Serialize, Debug))]
+#[derive(Encode, Decode, PartialEq, Eq, Clone)]
+pub struct PreRuntime<T>(pub [u8; 4], pub Vec<u8>, pub PhantomData<T>);
+
 /// Provides the slot duration inherent data for `Aura`.
 #[cfg(feature = "std")]
 pub struct InherentDataProvider {
```

File: srml/aura/src/lib.rs [hunk 2]
```diff
@@ -55,16 +55,21 @@ use srml_support::storage::StorageValue;
...
 use inherents::{RuntimeString, InherentIdentifier, InherentData, ProvideInherent, MakeFatalError};
 #[cfg(feature = "std")]
 use inherents::{InherentDataProviders, ProvideInherentData};
+#[cfg(feature = "std")]
+use serde::Serialize;
 
 mod mock;
 mod tests;
 
+pub use aura_primitives::AURA_ENGINE_ID;
```
Comment (marcio-diaz) [inline srml/aura/src/lib.rs]: Not needed?

File: node/runtime/src/lib.rs [hunk 3]
```diff
@@ -207,14 +207,16 @@ impl finality_tracker::Trait for Runtime {
 	type OnFinalizationStalled = grandpa::SyncedAuthorities<Runtime>;
 }
 
+use aura::AURA_ENGINE_ID;
```
Comment (marcio-diaz) [inline node/runtime/src/lib.rs]: Not needed?

File: core/offchain/src/testing.rs [hunk 4]
```diff
@@ -241,4 +241,3 @@ impl offchain::Externalities for TestOffchainExt {
 		}
 	}
 }
-
```
Comment (marcio-diaz) [inline core/offchain/src/testing.rs]: Add line

File: scripts/common.sh [hunk 5]
```diff
@@ -1,6 +1,7 @@
 #!/usr/bin/env bash
 
 ROOT=`dirname "$0"`
+unset RUSTC_WRAPPER
```
Comment (tomusdrw) [inline scripts/common.sh]: What's the purpose of this? Could you add a comment?

File: srml/aura/Cargo.toml [hunk 6]
```diff
@@ -16,11 +16,17 @@ timestamp = { package = "srml-timestamp", path = "../timestamp", default-feature
 staking = { package = "srml-staking", path = "../staking", default-features = false }
 session = { package = "srml-session", path = "../session", default-features = false }
 
+[dependencies.aura-primitives]
```
Comment (tomusdrw) [inline srml/aura/Cargo.toml]: Why not use the regular syntax we use everywhere? Is `version` actually needed?

File: core/sr-primitives/src/lib.rs [hunk 7]
```diff
@@ -688,22 +749,28 @@ macro_rules! impl_outer_log {
...
-			/// Converts `generic::DigestItem` into `$name`. If `generic::DigestItem` represents
-			/// a system item which is supported by the runtime, it is returned.
-			/// Otherwise we expect a `Other` log item. Trying to convert from anything other
-			/// will lead to panic in runtime, since the runtime does not supports this 'system'
-			/// log item.
+			/// Converts `generic::DigestItem` into `$name`. If
+			/// `generic::DigestItem` represents a system item which is
+			/// supported by the runtime, it is returned. Otherwise we expect a
+			/// `Other`, `PreDigest`, or `Consensus` log item. Trying to convert
+			/// from anything else will lead to panic at runtime, since the
```
Comment (marcio-diaz) [inline core/sr-primitives/src/lib.rs]: Why it doesn't support  `DigestItem::Seal`?

File: scripts/common.sh (hunk 5 above)
Comment (DemiMarie-parity) [reply scripts/common.sh]: I was trying to figure out why `cargo run --dev` was failing, and I wondered if `sccache` was failing to recompile something it needed to.   I don’t think it was, though.

File: srml/aura/Cargo.toml (hunk 6 above)
Comment (DemiMarie-parity) [reply srml/aura/Cargo.toml]: `version` is for if we want to publish on `crates.io`.  The different syntax is to avoid CI complaining about line length.

File: core/offchain/src/testing.rs (hunk 4 above)
Comment (DemiMarie-parity) [reply core/offchain/src/testing.rs]: The standard git commit hook complained about a blank line at EOF during a merge, but I can skip that check.

File: srml/aura/Cargo.toml (hunk 6 above)
Comment (tomusdrw) [reply srml/aura/Cargo.toml]: None of the other internal crates have `version` here, so we would need to rework this module anyway before publish. I don't think we should introduce `version` just yet.

I think you can break the regular syntax into multiple lines without issues:
```toml
aura-primitives = {
  package = "substrate-consensus-aura-primitives",
  path = "../../...",
  default-features = false,
}
```

File: core/sr-primitives/src/lib.rs (hunk 7 above)
Comment (DemiMarie-parity) [reply core/sr-primitives/src/lib.rs]: Seals are always stripped by the consensus engine, so runtimes never see them.

File: core/sr-primitives/src/lib.rs (hunk 1 above)
Comment (thiolliere) [reply core/sr-primitives/src/lib.rs]: OK, this change is legit as Log cannot be non-generic, maybe we could in the future introduce non-generic Log but this would require some change and should be made in another PR.

File: core/sr-primitives/src/lib.rs [hunk 8]
```diff
@@ -688,22 +753,28 @@ macro_rules! impl_outer_log {
...
+						if let Some(s) = gen.as_other()
 						.and_then(|value| $crate::codec::Decode::decode(&mut &value[..]))
-						.map($name)
-						.expect("not allowed to fail in runtime"),
+						.map($name) {
+							s
+						} else {
+							panic!("Unhandled digest in runtime")
+						}
+					}
```
[the remark this answers is not in the document]
Comment (marcio-diaz) [reply core/sr-primitives/src/lib.rs]: What is protecting this from panicking?

File: srml/aura/Cargo.toml (hunk 6 above)
Comment (DemiMarie-parity) [reply srml/aura/Cargo.toml]: > None of the other internal crates have `version` here, so we would need to rework this module anyway before publish. I don't think we should introduce `version` just yet.
> 

Will fix, thanks!

> I think you can break the regular syntax into multiple lines without issues:
> 
> ```toml
> aura-primitives = {
>   package = "substrate-consensus-aura-primitives",
>   path = "../../...",
>   default-features = false,
> }
> ```

That isn’t valid TOML, although it probably should be.

File: core/sr-primitives/src/lib.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (DemiMarie-parity) [reply core/sr-primitives/src/lib.rs]: The reason for avoiding `expect` is so that the line number displayed in case of a panic points to the correct place, rather than somewhere in `libcore`.

**Comment by gavofyork** [comment]
tests fail?

File: core/sr-primitives/src/lib.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (gavofyork) [reply core/sr-primitives/src/lib.rs]: > What is protecting this from panicking?

^^^

[the remark this answers is not in the document]
Comment (DemiMarie-parity) [reply core/sr-primitives/src/lib.rs]: @gavofyork @marcio-diaz This can only panic if the runtime is bad and does not handle a digest that it receives.  This means the runtime is broken, which is fatal.

[the remark this answers is not in the document]
Comment (marcio-diaz) [reply core/sr-primitives/src/lib.rs]: Isn't it possible to receive a bad digest?

**Comment by DemiMarie-parity** [comment]
@gavofyork 

> tests fail?

I made a silly mistake that caused compilation errors.  Now fixed.

File: core/sr-primitives/src/lib.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (DemiMarie-parity) [reply core/sr-primitives/src/lib.rs]: @marcio-diaz The old code also `panic`ed here (though with a worse error message), so this is not a regression.  If the bad digest comes from a block created by another node, the block will (correctly) be rejected.  Otherwise, the bad digest had to come from the runtime itself or from the consensus engine.  Both are fatal errors.

File: srml/aura/Cargo.toml [hunk 9]
```diff
@@ -20,7 +20,7 @@ session = { package = "srml-session", path = "../session", default-features = fa
 lazy_static = "1.0"
 parking_lot = "0.7.1"
 substrate-primitives = { path = "../../core/primitives" }
-runtime_io = { package = "sr-io", path = "../../core/sr-io" }
+runtime_io = { package = "sr-io", path = "../../core/sr-io", features = ["wasm-nice-panic-message"] }
```
Comment (bkchr) [inline srml/aura/Cargo.toml]:
```suggestion
runtime_io = { package = "sr-io", path = "../../core/sr-io" }
```

Comment (bkchr) [reply srml/aura/Cargo.toml]: This should not be activated!

Comment (pepyakin) [inline srml/aura/Cargo.toml]: Should this actually be here?

File: core/sr-primitives/src/lib.rs [hunk 10]
```diff
@@ -688,22 +753,29 @@ macro_rules! impl_outer_log {
...
-						.map($name)
-						.expect("not allowed to fail in runtime"),
+					_ => {
+						if let Some(s) = gen.as_other()
+							.and_then(|value| $crate::codec::Decode::decode(&mut &value[..]))
+							.map($name) {
+							s
+						} else {
+							panic!("we only reach here if the runtime did not handle a digest; \
+								   runtimes are required to handle all digests they receive; qed")
```
Comment (bkchr) [inline core/sr-primitives/src/lib.rs]: Spaces

File: core/sr-primitives/src/generic/digest.rs [hunk 11]
```diff
@@ -74,13 +74,14 @@ pub enum DigestItem<Hash, AuthorityId, SealSignature> {
 	/// be generated by the native code of any consensus engine, but this is not
 	/// checked (yet).
 	Consensus(ConsensusEngineId, Vec<u8>),
-	/// Put a Seal on it.
+	/// Put a Seal on it.  This is only used by native code, and is never seen
```
Comment (pepyakin) [inline core/sr-primitives/src/generic/digest.rs]:
```suggestion
	/// Put a Seal on it. This is only used by native code, and is never seen
```

File: core/sr-primitives/src/generic/digest.rs [hunk 12]
```diff
@@ -110,7 +111,8 @@ pub enum DigestItemRef<'a, Hash: 'a, AuthorityId: 'a, SealSignature: 'a> {
 	/// be generated by the native code of any consensus engine, but this is not
 	/// checked (yet).
 	Consensus(&'a ConsensusEngineId, &'a Vec<u8>),
-	/// Put a Seal on it.
+	/// Put a Seal on it.  This is only used by native code, and is never seen
```
Comment (pepyakin) [inline core/sr-primitives/src/generic/digest.rs]:
```suggestion
	/// Put a Seal on it. This is only used by native code, and is never seen
```

File: core/sr-primitives/src/lib.rs (hunk 10 above)
Comment (bkchr) [reply core/sr-primitives/src/lib.rs]: + `)` should be on a new line.

File: srml/babe/src/lib.rs [hunk 13]
```diff
@@ -56,6 +58,20 @@ impl BabeInherentData for InherentData {
 	}
 }
 
+/// Logs in this module.
+pub type Log<T> = RawLog<T>;
```
Comment (pepyakin) [inline srml/babe/src/lib.rs]: I am just curious why do we need this type alias?

File: core/sr-primitives/src/lib.rs [hunk 14]
```diff
@@ -611,6 +613,69 @@ macro_rules! impl_outer_config {
...
+	($name:ident $engine_id:pat, $binder:pat) => {
+		$crate::generic::DigestItem::$name($binder)
+	};
+}
+
+#[macro_export]
+#[doc(hidden)]
+macro_rules! __parse_expr {
+	(PreRuntime $engine_id:expr, $module:ident $internal:ident $binder:expr) => {
+		$internal::$module($module::RawLog::PreRuntime($engine_id, $binder, $crate::rstd::marker::PhantomData))
```
Comment (bkchr) [inline core/sr-primitives/src/lib.rs]:
```suggestion
		$internal::$module($module::RawLog::PreRuntime($engine_id, $binder, Default::default()))
```

And then you don't need `pub use rstd`.

Same for below

File: srml/babe/src/lib.rs [hunk 15]
```diff
@@ -17,19 +17,21 @@
 //! Consensus extension module for BABE consensus.
 
 #![cfg_attr(not(feature = "std"), no_std)]
-#![forbid(unsafe_code, warnings)]
+#![forbid(unsafe_code)]
```
Comment (pepyakin) [inline srml/babe/src/lib.rs]: Is this change intended? If so, why?

File: core/sr-primitives/src/lib.rs [hunk 16]
```diff
@@ -688,22 +753,29 @@ macro_rules! impl_outer_log {
...
+						$name($crate::__parse_expr!($sitem b, $module $internal a)),
 					)*)*
-					_ => gen.as_other()
-						.and_then(|value| $crate::codec::Decode::decode(&mut &value[..]))
-						.map($name)
-						.expect("not allowed to fail in runtime"),
+					_ => {
+						if let Some(s) = gen.as_other()
+							.and_then(|value| $crate::codec::Decode::decode(&mut &value[..]))
+							.map($name) {
```
Comment (pepyakin) [inline core/sr-primitives/src/lib.rs]: I think this improves the legibility, feel free to ignore

```suggestion
							.map($name)
						{
```

File: srml/babe/src/lib.rs (hunk 13 above)
Comment (thiolliere) [reply srml/babe/src/lib.rs]: I think impl_outer_log macro use the RawLog definition. And this might be for type inference

File: node/runtime/src/lib.rs [hunk 17]
```diff
@@ -59,7 +59,7 @@ pub const VERSION: RuntimeVersion = RuntimeVersion {
 	impl_name: create_runtime_str!("substrate-node"),
 	authoring_version: 10,
 	spec_version: 90,
```
Comment (thiolliere) [inline node/runtime/src/lib.rs]: doesn't this change spec version ?

File: core/sr-primitives/src/lib.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (marcio-diaz) [reply core/sr-primitives/src/lib.rs]: > The old code also paniced here (though with a worse error message), so this is not a regression.

I know ^^.

>  If the bad digest comes from a block created by another node, the block will (correctly) be rejected

For curiosity, where is this logic? `check_header`? Because I'm not so sure in that case.

File: srml/babe/src/lib.rs (hunk 15 above)
Comment (DemiMarie-parity) [reply srml/babe/src/lib.rs]: Yes, to avoid breakage from future Rust versions.

File: srml/babe/src/lib.rs (hunk 13 above)
Comment (DemiMarie-parity) [reply srml/babe/src/lib.rs]: @pepyakin

> I am just curious why do we need this type alias?

`impl_outer_log!` requires it.

File: node/runtime/src/lib.rs (hunk 17 above)
Comment (DemiMarie-parity) [reply node/runtime/src/lib.rs]: I meant to bump `spec_version`, but it got lost in a merge conflict resolution.
